modules/cudaoptflow/include/opencv2/cudaoptflow.hpp

```diff
@@ -342,6 +386,56 @@ class CV_EXPORTS_W OpticalFlowDual_TVL1 : public DenseOpticalFlow
             bool useInitialFlow = false);
 };
 
+//
+// NvidiaOpticalFlow
+//
+
+/** @brief Class for computing the optical flow vectors between two images using NVIDIA Optical Flow hardware and Optical Flow SDK 1.0.
+@note
+- A sample application demonstrating the use of NVIDIA Optical Flow can be found at
+opencv_source_code/samples/gpu/nvidia_optical_flow.cpp
+- An example application comparing accuracy and performance of NVIDIA Optical Flow with other optical flow algorithms in OpenCV can be found at
+opencv_source_code/samples/gpu/optical_flow.cpp
+*/
+class CV_EXPORTS NvidiaOpticalFlow_1_0 : public NvidiaHWOpticalFlow
+{
+public:
+    /** @brief The NVIDIA optical flow hardware generates flow vectors at granularity gridSize, which can be queried via function getGridSize().
+    * Upsampler() helper function converts the hardware-generated flow vectors to dense representation (1 flow vector for each pixel)
+    * using nearest neighbour upsampling method.
+
+    @param flow Buffer of type CV_16FC2 containing flow vectors generated by calc().
+    @param width Width of the input image in pixels for which these flow vectors were generated.
+    @param height Height of the input image in pixels for which these flow vectors were generated.
+    @param gridSize Granularity of the optical flow vectors returned by calc() function. Can be queried using getGridSize().
+    @param upsampledFlow Buffer of type CV_32FC2, containing upsampled flow vectors, each flow vector for 1 pixel, in the pitch-linear layout.
+    */
+    CV_WRAP virtual void upSampler(InputArray flow, int width, int height,
+        int gridSize, InputOutputArray upsampledFlow) = 0;
+
+    /** @brief Instantiate NVIDIA Optical Flow
+
+    @param width Width of input image in pixels.
+    @param height Height of input image in pixels.
+    @param perfPreset Optional parameter. Refer [NV OF SDK documentation](https://developer.nvidia.com/opticalflow-sdk) for details about presets.
+                      Defaults to NV_OF_PERF_LEVEL_SLOW.
+    @param enableTemporalHints Optional parameter. Flag to enable temporal hints. When set to true, the hardware uses the flow vectors
+                               generated in previous call to calc() as internal hints for the current call to calc().
+                               Useful when computing flow vectors between successive video frames. Defaults to false.
+    @param enableExternalHints Optional Parameter. Flag to enable passing external hints buffer to calc(). Defaults to false.
+    @param enableCostBuffer Optional Parameter. Flag to enable cost buffer output from calc(). Defaults to false.
+    @param gpuId Optional parameter to select the GPU ID on which the optical flow should be computed. Useful in multi-GPU systems. Defaults to 0.
+    */
+    CV_WRAP static Ptr<NvidiaOpticalFlow_1_0> create(
```

Comment:
will this call raise exception or return empty pointer if specified `gpuId` H/W doesn't support this feature? (please clarify this case in documentation)

(also we probably need some check for that in tests)

Reply:
So if specific GPU is not compatible then HMODULE hModule = LoadLibrary(MODULENAME); wont be able to get the library file or dll and it will return appropriate error..

Reply:
There is `gpuId`.
2 gpus in system. They are different.

Is it clear now why approach with `LoadLibrary` will not work?

Reply:
I got your point. In that case nvCreateOpticalFlowCuda() will fail and it will return exception. It wont create empty pointer. Il try out this situation and add appropriate checks in tests.

Reply:
I added try catch block so that test suite wont crash and stop. So if specified HW is not supported test will fail and return silently. Will this be Okay?

Reply:
Silent errors is a bad practice.
Exceptions is more useful for users (at least they have error message, and sometimes stack trace).

Looks like, there is no way to detect support during `create()` function.
Checking API support is still actual for tests (to skip test on unsupported H/W instead of failures).

Reply:
I meant it wont crash abruptly now rather it will give the exception and exit.
OpenCV uses gtest framework right? So cant we put restriction that particular tests should run only on turing and above gpu archs?

Reply:
OpenCV uses patched Google Tests framework.

You can use this code to "skip" test.
```
if (<not_supported_configuration_condition>)
    throw SkipTestException("Current configuration is not supported");
```
Main question here is "condition" implementation.

---

Alternative (exception variant):
```
cv::Ptr<cv::cuda::NvidiaOpticalFlow_1_0> d_nvof;
try
{
    d_nvof =cv::cuda::NvidiaOpticalFlow_1_0::create(...);
}
catch (... not supported exception ...)
{
    throw SkipTestException("Current configuration is not supported");
}
```

Reply:
I seperated error code for each specific error to define that condition. We mapped opencv error to as much similar error can be to nvof error.